extension/server/src/providers/codeActions.ts

```diff
@@ -48,32 +58,282 @@ export default async function genericCodeActionsProvider(params: CodeActionParam
+		if (workspaceFolder) {
+			// Build new test file uri
+			const parsedPath = path.parse(document.uri);
+			const fileName = parsedPath.base;
+			const testFileName = `${parsedPath.name}.test${parsedPath.ext}`;
+			const testFileUri = workspaceFolder ?
+				`${workspaceFolder.uri}/qtestsrc/${testFileName}` :
+				`${parsedPath.dir}/${testFileName}`;
+
+			// Test case generation
+			const currentProcedure = exportProcedures.find(sub => sub.range.start && sub.range.end && range.start.line >= sub.range.start && range.end.line <= sub.range.end);
+			if (currentProcedure) {
+				const testCaseSpec = await getTestCaseSpec(docs, currentProcedure, workspaceFolder);
+				const newTestSuite = generateTestSuite([testCaseSpec]);
+				const testCaseAction = CodeAction.create(`Generate test case for '${currentProcedure.name}'`, CodeActionKind.RefactorExtract);
+				testCaseAction.edit = {
+					documentChanges: [
+						CreateFile.create(testFileUri, { ignoreIfExists: true }),
+						TextDocumentEdit.create({ uri: testFileUri, version: null }, [TextEdit.insert(Position.create(0, 0), newTestSuite.join(`\n`))])
+					]
+				};
+				codeActions.push(testCaseAction);
+			}
 
-		const refactorAction = CodeAction.create(`Create IBM i test case`, CodeActionKind.RefactorExtract);
+			// Test suite generation
+			const newTestCases = await Promise.all(exportProcedures.map(async proc => await getTestCaseSpec(docs, proc, workspaceFolder)));
+			const newTestSuite = generateTestSuite(newTestCases);
+			const testSuiteAction = CodeAction.create(`Generate test suite for '${fileName}'`, CodeActionKind.RefactorExtract);
+			testSuiteAction.edit = {
+				documentChanges: [
+					CreateFile.create(testFileUri, { ignoreIfExists: true }),
+					TextDocumentEdit.create({ uri: testFileUri, version: null }, [TextEdit.insert(Position.create(0, 0), newTestSuite.join(`\n`))])
+				]
+			};
+			codeActions.push(testSuiteAction);
+		}
+	}
 
-		refactorAction.edit = {
-			changes: {
-				['mynewtest.rpgle']: [
-					TextEdit.insert(
-						Position.create(0, 0), // Insert at the start of the new test case file
-						[
-							`**free`,
-							``,
-							`dcl-proc test_${currentProcedure.name.toLowerCase()} export;`,
-							``,
-							`end-proc;`
-						].join(`\n`)
+	return codeActions;
+}
 
-					)
-				]
-			},
-		};
 
-		return refactorAction;
+function generateTestSuite(testCaseSpecs: TestCaseSpec[]) {
+	const prototypes = testCaseSpecs.map(tc => tc.prototype.length > 0 ? [``, ...tc.prototype] : tc.prototype).flat();
+	const testCases = testCaseSpecs.map(tc => tc.testCase.length > 0 ? [``, ...tc.testCase] : tc.testCase).flat();
+	const allIncludes = testCaseSpecs.map(tc => tc.includes).flat();
+	const uniqueIncludes = [...new Set(allIncludes)];
+
+	return [
+		`**free`,
+		``,
+		`ctl-opt nomain;`,
+		...prototypes,
+		``,
+		`/include qinclude,TESTCASE`,
+		...uniqueIncludes,
```

Comment:
I think some logic needs to happen for `uniqueIncludes`.

If `uniqueInclude` is a POSIX path (IFS/local), then please use single quotes around the `/copy`/`/include` path:

```
/include 'qrpgleref/empdet.rpgleinc'
```

And to confirm, members should use this format:

```
/include qrpgleref,empdet
```